components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -0,0 +1,496 @@
+        cmd.mut_header().set_peer(self.peer().clone());
+        let proposal_size = cmd.compute_size();
+        if proposal_size as usize > size_limit {
+            return Err(box_err!(
+                "pessimistic locks size {} exceed size limit {}, skip merging.",
+                proposal_size,
+                size_limit
+            ));
+        }
+
+        {
+            let mut pessimistic_locks = RwLockUpgradableReadGuard::upgrade(pessimistic_locks);
+            pessimistic_locks.status = LocksStatus::MergingRegion;
+        }
+        debug!(
+            self.logger,
+            "propose {} pessimistic locks before prepare merge",
+            cmd.get_requests().len();
+        );
+        self.propose(store_ctx, cmd.write_to_bytes().unwrap())?;
+        Ok(())
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    // Match v1::exec_prepare_merge.
+    pub fn apply_prepare_merge(
+        &mut self,
+        req: &AdminRequest,
+        log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        PEER_ADMIN_CMD_COUNTER.prepare_merge.all.inc();
+
+        let prepare_merge = req.get_prepare_merge();
+        let index = prepare_merge.get_min_index();
+        // Note: the check against first_index is removed in v2.
+        let mut region = self.region().clone();
+        let region_version = region.get_region_epoch().get_version() + 1;
+        region.mut_region_epoch().set_version(region_version);
+        // In theory conf version should not be increased when executing prepare_merge.
+        // However, we don't want to do conf change after prepare_merge is committed.
+        // This can also be done by iterating all proposal to find if prepare_merge is
+        // proposed before proposing conf change, but it make things complicated.
+        // Another way is make conf change also check region version, but this is not
+        // backward compatible.
+        let conf_version = region.get_region_epoch().get_conf_ver() + 1;
+        region.mut_region_epoch().set_conf_ver(conf_version);
+        let mut merging_state = MergeState::default();
+        merging_state.set_min_index(index);
+        merging_state.set_target(prepare_merge.get_target().to_owned());
+        merging_state.set_commit(log_index);
+
+        self.region_state_mut().set_region(region.clone());
+        self.region_state_mut().set_state(PeerState::Merging);
+        self.region_state_mut()
+            .set_merge_state(merging_state.clone());
+
+        PEER_ADMIN_CMD_COUNTER.prepare_merge.success.inc();
+
+        Ok((
+            AdminResponse::default(),
+            AdminCmdResult::PrepareMerge(PrepareMergeResult {
+                region_state: self.region_state().clone(),
+                state: merging_state,
+            }),
+        ))
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    // Match v1::on_ready_prepare_merge.
+    pub fn on_apply_res_prepare_merge<T: Transport>(
```

Comment:
Why need `Transport` constraint?

Reply:
`update_merge_progress_on_apply_res_prepare_merge` triggers merge tick immediately. Merge tick might rollback, which sends an ExtraMessage to leader.

Reply:
How about notifying sibling peer directly? So the prepare state is stored in sibling peer, and sibling peer is responsible to propose commit merge in timely fashion. We don't need a tick to make it work.

Reply:
The commit merge command needs source peer logs (the whole CatchUpLogs thing). And the prepare state still needs to be maintained in source peer to reject further writes.

Reply:
> The commit merge command needs source peer logs

That seems to make maintaining pending merge in target region more beneficial as  it can reduce unnecessary read. But it's a complicated change, may do it later.